### PR TITLE
[WIP] cudnn v8 fmha integration

### PR DESCRIPTION
**Pull request by Wong4j**
### PR types
New features

### PR changes
OPs

### Description
CUDNN v8 provide fused MHA kernel to accelerate MultiHeadAttention(refer to [cudnn-frontend/fused_mha_sample](https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fused_mha_sample.cpp)). To be exact, it only fuse the calculation of _**scaled dot product attention**_, which does not include **_qkv projection_** and **_output projection_**, but only includes **_scale, bmm1, mask, softmax, dropout, and bmm2_**. So, let's call it **_fused_dot_product_attention_** instead of **_fused_multi_head_attention_**.

1. Added the following OPs to compute **_scaled dot product attention_**  (forward + backward):
    - fused_dot_product_self_attention/fused_dot_product_self_attention_grad
    - fused_dot_product_cross_attention/fused_dot_product_cross_attention_grad
2. Added 2 optional arguments in paddle.nn.MultiHeadAttention: **_fuse_qkv_** and **_mode_**
    - Default: **_fuse_qkv_**==False and **_mode_** == None,  the QKV projection will use 3 different weight tensor.
    - If **_fuse_qkv_**==True and **_mode_** == "self_attn", the QKV projection will use a single weight tensor with shape [embed_dim, 3*embed_dim], and the output of QKV projection can be reshaped into [b, s, 3, h, d].
    - If **_fuse_qkv_**==True and **_mode_** == "cross_attn", the KV projection will use a single weight tensor with shape [embed_dim, 2*embed_dim], and the output of KV projection can be reshaped into [b, s, 2, h, d].
3. Added a new IR pass to detect above "self_attn" pattern and "cross_attn" pattern. 
4. Added a class member fuse_dot_product_attention to BuildStrategy to enable fuse_dot_product_attention_pass.

**Note:** 
The difference between the two attention mode is that the input of self-attention requires QKV to be a single buffer, while the input of cross-attention only requires KV to be a single buffer, with Q existing separately in another buffer.   
More specific details are that the **_QKV_** input for **_fused_dot_product_self_attention_** op has shape **_[b, s, 3, h, d]_**, where _**[:, :, 0, :, :]**_ for **_Q_**,  **_[:, :, 1, :, :]_** for **_K_**,  **_[:, :, 2, :, :]_** for **_V_**. The KV input for **_fused_dot_product_cross_attention_** op has shape **_[b, s, 2, h, d]_**, where **_[:, :, 0, :, :]_** for **_K_**,  **_[:, :, 1, :, :]_** for **_V_**.

**Usage sample (CUDNN >= 8.9.1 is required):**
```
class TransformerLayer(paddle.nn.Layer):
    def __init__(self, hidden, num_heads):
        super().__init__()
        self.encoder_layer = paddle.nn.TransformerEncoderLayer(hidden, num_heads, 4*hidden, dropout=0.1, fuse_qkv=True)
        self.encoder = paddle.nn.TransformerEncoder(self.encoder_layer, 2)

    def forward(self, q, mask):
        out = self.encoder(q, mask)
        loss = paddle.mean(out)
        return loss

paddle.enable_static()
main_prog = paddle.static.Program()
startup_prog = paddle.static.Program()

batch_size = 32
head_size = 64
num_head = 12
hidden = head_size * num_heads
seq_len = 128

with paddle.static.program_guard(main_prog, startup_prog):
    enc_input = paddle.static.data(name='enc_input', shape=[-1, -1, hidden], dtype="float32")
    attn_mask = paddle.static.data(name='mask', shape=[-1, 1, 1, seq_len], dtype="int32")
    model = TransformerLayer(hidden, num_heads)
    loss = model(enc_input, attn_mask)
    opt = paddle.optimizer.SGD(learning_rate=0.1)
    amp_list = paddle.static.amp.CustomOpLists(custom_white_list=['softmax'])
    opt = paddle.static.amp.decorate(
        optimizer=opt,
        amp_lists=amp_list,
        init_loss_scaling=128.0,
        use_dynamic_loss_scaling=True)
    opt.minimize(loss)


place = paddle.CUDAPlace(0)
exe = paddle.static.Executor(place)
exe.run(startup_prog)

build_strategy = paddle.static.BuildStrategy()
build_strategy.fuse_dot_product_attention = True
main_prog = paddle.static.CompiledProgram(main_prog)
main_prog = main_prog.with_data_parallel(loss_name=loss.name,
                                         build_strategy=build_strategy,
                                         places=place)
exe.run(main_prog, feed=feed)
# 2 subgraphs be fused to fused_dot_product_self_attention op
# 2 subgraphs be fused to fused_dot_product_self_attention_grad op
```

BERT Perf test (8 x A100 80GB):
| seqlen | original (sequence per second) |  cuDNN MHA (sequence per second) | speedup|
| ------- | ------- | ------- | -------|
| 128     |   610 |   685 |  1.12x |
| 512     |   116 |   157 |  1.35x |

BERT convergence test (8 x A100 80GB x 32 nodes):
- Phase1 final loss: 1.34
- Phase2 final loss: 1.24
- SQuAD F1: 90.5

BERT code:
[https://github.com/NVIDIA/DeepLearningExamples/tree/master/PaddlePaddle/LanguageModeling/BERT](https://github.com/NVIDIA/DeepLearningExamples/tree/master/PaddlePaddle/LanguageModeling/BERT)

### CODE REVIEW

**Comment by jeng1220** [comment]
這個 PR 和 #50196 都涉及 Paddle 和 cuDNN Front-end 的整合，麻煩優先審核、合併 #50196，這樣 merge conflict 會比較少

**Comment by Wong4j** [comment]
# IR pass
IR pass部分的实现较为复杂，这里做一些解释以便review

## 设计思路:

**QKV输入**：fused_dot_product_attention 不计算qkv projection和out projection，只计算中间的scale, bmm1, mask, softmax, dropout, bmm2等。并且要求输入的qkv(self attention)或者kv(cross attention)是同一块buffer。但nn.MultiHeadAttention中的qkv projection的输出是三个独立的q, k, v tensor，不能直接作为fused_dot_product_attention的输入。因此我们对nn.MultiHeadAttention新增了两个optional的参数：fuse_qkv和mode。

1. 若fuse_qkv=False, mode=None, 将按照paddle默认的实现，将qkv分开计算
2. 若fuse_qkv=True, mode=”self_attn”, qkv projection的weight合成一整块，shape为[hidden_dim, 3*hidden_dim]来计算，输出shape为[b, s, 3*hidden_dim]，reshape成[b, s, 3, h, d]，作为fused_dot_product_self_attention的输入。
3. 若fuse_qkv=True, mode=”cross_atten”, kv projection的weight合成一整块，q projection单独计算。则输出reshape成两个：q→[b, s, h, d]  kv→[b, s, 2, h, d], 作为fused_dot_product_cross_attention的输入。

由于我们的设计是用IR pass来对subgraph进行遍历和替换，并没有直接在nn.MultiHeadAttention的forward中调用fused op，所以融合的qkv或kv会被再次slice成独立的tensor，然后来按paddle native mha的流程来计算。如下左图。

仅当用户开启build_strategy.fuse_dot_product_attention=True（并且系统满足cudnn版本等要求）之后，这些slice op连同后面的dot_product_attention的subgraph会被替换成fused op。如下右图。

image


注意: 由于fused_dot_product_attention op并不能直接输入mask，而是输入actual_seqlen，所以需要插入额外的squeeze, reduce_sum等op将mask转化为actual_seqlen。

## 实现细节：

**fuse_dot_product_attention_pass.h:** 
**fuse_dot_product_attention_pass.cc**

1. **3个OpCache**
    1. SequenceMetaCache: 用来存actual_seqlen节点（等同于mask），比如BERT large有24个layer，对同一个输入batch来说，计算24次mha的fwd/bwd都会使用同一组mask，所以只需要计算一次actual seqlen然后存在cache中。
    2. SoftmaxOutputCache：cudnn mha中bwd的计算需要fwd的中间结果，即softmax的输出（若有dropout，则为dropout的输出）。所以fwd计算完softmax，存储在cache中，然后bwd再取出来用。
    3. QKVCache：同理，fwd计算时将qkv存储在cache中，在对应的bwd计算中取出来，用来计算dq, dk, dv。
2. **nodes_to_remove**
    
    如上所述，以BERT-large为例，需要将24个fwd/bwd中所有的subgraph都替换成fused op。由于某些node的指针存储在cache中，在bwd时要再次使用，所以不能每次替换一个subgraph后就立即删除掉被替换的节点，而是替换完所有的subgraph后再删除。
    
3. **InsertActualSeqlenOp_**
    
    插入一些小的paddle op将mask转化成q_actual seqlen和k_actual_seqlen，目前支持两种情况：
    
    1. mask shape: [b, 1, 1, s], BERT encoder中常见的shape，需要转化成q_actual_seqlen和k_actual_seqlen，首先插入**squeeze**，去掉中间的两个1维度变成[b, s]，然后用ones_like拷贝一个全为1的[b, s] tensor。对这两个tensor分别做reduce_sum就可以得到shape为[b, ]的q_actual_seqlen和k_actual_seqlen。   
     这种情况下插入了squeeze, ones_like, reduce_sum, reduce_sum 4个op。实际上这种mask等效于只在k上做了mask，所以q_actual_seqlen的值都是max_seqlen 
        
    2. mask shape: [b, 1, s_q, s_kv], decoder中常用的shape。类似的，等效于q和k都需要mask，需要插入slice, slice, reduce_sum, reduce_sum 4个op。
    3. 未来若有需要，应当支持causal masking，暂未实现

**Comment by JamesLim-sy** [comment]
> # IR pass
> 
> IR pass部分的实现较为复杂，这里做一些解释以便review
> ## 设计思路:
> 
> **QKV输入**：fused_dot_product_attention 不计算qkv projection和out projection，只计算中间的scale, bmm1, mask, softmax, dropout, bmm2等。并且要求输入的qkv(self attention)或者kv(cross attention)是同一块buffer。但nn.MultiHeadAttention中的qkv projection的输出是三个独立的q, k, v tensor，不能直接作为fused_dot_product_attention的输入。因此我们对nn.MultiHeadAttention新增了两个optional的参数：fuse_qkv和mode。
> 
>     1. 若fuse_qkv=False, mode=None, 将按照paddle默认的实现，将qkv分开计算
> 
>     2. 若fuse_qkv=True, mode=”self_attn”, qkv projection的weight合成一整块，shape为[hidden_dim, 3_hidden_dim]来计算，输出shape为[b, s, 3_hidden_dim]，reshape成[b, s, 3, h, d]，作为fused_dot_product_self_attention的输入。
> 
>     3. 若fuse_qkv=True, mode=”cross_atten”, kv projection的weight合成一整块，q projection单独计算。则输出reshape成两个：q→[b, s, h, d]  kv→[b, s, 2, h, d], 作为fused_dot_product_cross_attention的输入。
> 
> 
> 由于我们的设计是用IR pass来对subgraph进行遍历和替换，并没有直接在nn.MultiHeadAttention的forward中调用fused op，所以融合的qkv或kv会被再次slice成独立的tensor，然后来按paddle native mha的流程来计算。如下左图。
> 
> 仅当用户开启build_strategy.fuse_dot_product_attention=True（并且系统满足cudnn版本等要求）之后，这些slice op连同后面的dot_product_attention的subgraph会被替换成fused op。如下右图。
> 
> image
> 
> 注意: 由于fused_dot_product_attention op并不能直接输入mask，而是输入actual_seqlen，所以需要插入额外的squeeze, reduce_sum等op将mask转化为actual_seqlen。
> ## 实现细节：
> 
> **fuse_dot_product_attention_pass.h:** **fuse_dot_product_attention_pass.cc**
> 
>     1. **3个OpCache**
>        
>        1. SequenceMetaCache: 用来存actual_seqlen节点（等同于mask），比如BERT large有24个layer，对同一个输入batch来说，计算24次mha的fwd/bwd都会使用同一组mask，所以只需要计算一次actual seqlen然后存在cache中。
>        2. SoftmaxOutputCache：cudnn mha中bwd的计算需要fwd的中间结果，即softmax的输出（若有dropout，则为dropout的输出）。所以fwd计算完softmax，存储在cache中，然后bwd再取出来用。
>        3. QKVCache：同理，fwd计算时将qkv存储在cache中，在对应的bwd计算中取出来，用来计算dq, dk, dv。
> 
>     2. **nodes_to_remove**
>        如上所述，以BERT-large为例，需要将24个fwd/bwd中所有的subgraph都替换成fused op。由于某些node的指针存储在cache中，在bwd时要再次使用，所以不能每次替换一个subgraph后就立即删除掉被替换的节点，而是替换完所有的subgraph后再删除。
> 
>     3. **InsertActualSeqlenOp_**
>        插入一些小的paddle op将mask转化成q_actual seqlen和k_actual_seqlen，目前支持两种情况：
>        
>        1. mask shape: [b, 1, 1, s], BERT encoder中常见的shape，需要转化成q_actual_seqlen和k_actual_seqlen，首先插入**squeeze**，去掉中间的两个1维度变成[b, s]，然后用ones_like拷贝一个全为1的[b, s] tensor。对这两个tensor分别做reduce_sum就可以得到shape为[b, ]的q_actual_seqlen和k_actual_seqlen。
>           这种情况下插入了squeeze, ones_like, reduce_sum, reduce_sum 4个op。实际上这种mask等效于只在k上做了mask，所以q_actual_seqlen的值都是max_seqlen
>        2. mask shape: [b, 1, s_q, s_kv], decoder中常用的shape。类似的，等效于q和k都需要mask，需要插入slice, slice, reduce_sum, reduce_sum 4个op。
>        3. 未来若有需要，应当支持causal masking，暂未实现

请问下。这块存储的3个OpCache，是否与`flash_attention`中的`softmax_lse`, `fmha_out`, `seq_len_r` 的三种前向计算的驻留信息相仿。

**Comment by Wong4j** [comment]
> 请问下。这块存储的3个OpCache，是否与`flash_attention`中的`softmax_lse`, `fmha_out`, `seq_len_r` 的三种前向计算的驻留信息相仿。

我目前还不熟悉flash attention的细节。参考cudnn frontend 的 sample： [fused_mha_sample.cpp](https://github.com/NVIDIA/cudnn-frontend/blob/v0.8/samples/fused_mha_sample.cpp)

- actual_seqlen很好理解，比如
```
# first batch.
shape: [2, 8]
mask: [[1, 1, 1, 1, 1, 1, 1, 0],
       [1, 1, 1, 1, 1, 0, 0, 0]]
# second batch
shape: [2, 16]
mask: [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0],
       [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0]]
```
第一个batch的 max_seqlen = 8, actual seqlen = [7, 5]
第二个batch的 max_seqlen = 16, actual seqlen = [13, 8]

- SoftmaxOutputCache
forward计算出来的中间结果`S`，在backward需要作为输入，如下图
image

- QKVCache
如上图，forward需要Q, K, V作为输入，backward也需要Q, K, V作为输入，用来计算dQ, dK, dV。所以在计算forward的时候把QKV存到cache中，backward可以直接取出来用。

**Comment by Wong4j** [comment]
I have submit a new PR https://github.com/PaddlePaddle/Paddle/pull/58680. Close this PR
